### PR TITLE
Univariate metrics based on sparse union

### PR DESCRIPTION
**Pull request by lquerel**
This PR implements the encoding and decoding of univariate metrics based on Arrow union.

Gauge, Sum, and Summary encoding and decoding are implemented and tested.

Things that will be implemented in next PR:
- https://github.com/f5/otel-arrow-adapter/issues/38
- https://github.com/f5/otel-arrow-adapter/issues/36
- https://github.com/f5/otel-arrow-adapter/issues/39

### CODE REVIEW

**Comment by lquerel** [comment]
This PR implements the encoding and decoding of univariate metrics based on Arrow union.
Gauge, Sum, and Summary encoding and decoding are supported.

Things that will be implemented in next PR:
- Histogram and Exponential Histogram encoding/decoding
- Multivariate metrics support
- Integration with benchmark tool

File: pkg/otel/metrics/arrow/all_test.go [hunk 1]
```diff
@@ -0,0 +1,668 @@
...
...pan_id":"qgAAAAAAAAA=","time_unix_nano":1,"trace_id":"qgAAAAAAAAAAAAAAAAAAAA==","value":[1,1.5]}],"flags":3,"start_time_unix_nano":3,"time_unix_nano":4,"value":[0,3]}],"is_monotonic":true}],"description":"sum-2-desc","name":"sum-2","unit":"sum-2-unit"},{"data":[2,{"data_points":[{"attributes":[{"key":"str","value":[0,"string1"]},{"key":"int","value":[1,1]},{"key":"double","value":[2,1]},{"key":"bool","value":[3,true]},{"key":"bytes","value":[4,"Ynl0ZXMx"]}],"count":1,"flags":1,"quantile":[{"quantile":0.1,"value":1.5},{"quantile":0.2,"value":2.5}],"start_time_unix_nano":1,"sum":1.5,"time_unix_nano":2},{"attributes":[{"key":"str","value":[0,"string2"]},{"key":"int","value":[1,2]},{"key":"double","value":[2,2]},{"key":"bytes","value":[4,"Ynl0ZXMy"]}],"count":2,"flags":2,"quantile":[{"quantile":0.2,"value":2.5}],"start_time_unix_nano":3,"sum":2.5,"time_unix_nano":4}]}],"description":"summary-3-desc","name":"summary-3","unit":"summary-3-unit"}],"schema_url":"schema-2","scope":{"attributes":[{"key":"str","value":[0,"string2"]},{"key":"int","value":[1,2]},{"key":"double","value":[2,2]},{"key":"bytes","value":[4,"Ynl0ZXMy"]}],"dropped_attributes_count":1,"name":"scope2","version":"1.0.2"}}
```
Comment (jmacd) [inline pkg/otel/metrics/arrow/all_test.go:364]: This test worries me -- as it's really a "change detector" but doesn't build in a correctness check. There are probably ways we can make this more readable or maintainable, but they're OK for temporary purposes.

**Review by jmacd — COMMENTED** [review]
(Note: I've only just begun this review & have to stop for the day.)

File: pkg/otel/metrics/arrow/all_test.go (hunk 1 above)
Comment (lquerel) [reply pkg/otel/metrics/arrow/all_test.go:364]: I agree and am open to suggestions.

Note that there is also a series of validation_test.go files using another approach (based on logical comparison).

File: pkg/datagen/metrics.go [hunk 2]
```diff
@@ -134,9 +133,8 @@ func (dg *DataGenerator) SystemMemoryUsage(metric pmetric.Metric) {
 	metric.SetDescription("Bytes of memory in use.")
 	metric.SetUnit("By")
 	sum := metric.SetEmptySum()
-	// TODO
-	// sum.SetAggregationTemporality(pmetric.MetricAggregationTemporalityCumulative)
-	// sum.SetIsMonotonic(false)
+	//sum.SetAggregationTemporality(pmetric.AggregationTemporalityCumulative)
+	//sum.SetIsMonotonic(false)
```
Comment (jmacd) [inline pkg/datagen/metrics.go:137]: Should this file have real settings for temporality and monotonicity?

File: pkg/otel/common/arrow/attributes.go [hunk 3]
```diff
@@ -76,13 +76,13 @@ func (b *AttributesBuilder) Append(attrs pcommon.Map) error {
 
 	var err error
 	attrs.Range(func(key string, v pcommon.Value) bool {
-		// Append the key
+		// AppendNumberDataPointValue the key
```
Comment (jmacd) [inline pkg/otel/common/arrow/attributes.go:79]:
```suggestion
		// Append the key
```

File: pkg/otel/common/arrow/attributes.go [hunk 4]
```diff
@@ -76,13 +76,13 @@ func (b *AttributesBuilder) Append(attrs pcommon.Map) error {
...
 		err := b.kb.AppendString(key)
 		if err != nil {
 			return false
 		}
 
-		// Append the value
+		// AppendNumberDataPointValue the value
```
Comment (jmacd) [inline pkg/otel/common/arrow/attributes.go:85]:
```suggestion
		// Append the value
```

File: pkg/otel/metrics/arrow/metric_set.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+		arrow.Field{Name: constants.DATA, Type: UnivariateMetricDT},
+	)
+
+	// MultivariateMetricsDT is the Arrow Data Type describing a set of multivariate metrics.
+	// Multivariate metrics are metrics sharing the same attributes, start time, and end time.
+	MultivariateMetricsDT = arrow.StructOf([]arrow.Field{
+		{Name: constants.NAME, Type: acommon.DictU16String},
+		{Name: constants.DESCRIPTION, Type: acommon.DictU16String},
+		{Name: constants.UNIT, Type: acommon.DictU16String},
+		// attributes
```
Comment (jmacd) [inline pkg/otel/metrics/arrow/metric_set.go:30]:
```suggestion
		// TODO
		// attributes
```

File: pkg/otel/metrics/otlp/metrics.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+	metricsIds, err := SchemaToIds(record.Schema())
+	if err != nil {
+		return metrics, err
+	}
+
+	resMetricsSlice := metrics.ResourceMetrics()
+	resSpansCount := int(record.NumRows())
+	resMetricsSlice.EnsureCapacity(resSpansCount)
+
+	// TODO there is probably two nested lists that could be replaced by a single list (metrics, resource spans). This could simplify a future query layer.
```
Comment (jmacd) [inline pkg/otel/metrics/otlp/metrics.go:25]: Curious about this -- are you referring to a mix of OTLP logs/metrics/spans going through Arrow w/ a single Resource?

**Review by jmacd — APPROVED** [review]

File: pkg/datagen/metrics.go (hunk 2 above)
Comment (lquerel) [reply pkg/datagen/metrics.go:137]: Yes but I need to change some unit tests first.

File: pkg/otel/metrics/otlp/metrics.go (hunk 6 above)
Comment (lquerel) [reply pkg/otel/metrics/otlp/metrics.go:25]: It's not clear. I need to revisit that.

**Pull request merged**